design-guidelines/form-ux.md

```diff
@@ -0,0 +1,189 @@
+# Form Design Guidelines
+
+A form may be used to collect data that is required to complete a task.
+
+## Form contexts
+
+The context in which a form is rendered depends on the complexity of the form and how critical the form's purpose is to the rest of the app.
+
+### Full page
+
+A form may be rendered on it's own page helps users to focus on completing their task with minimal distractions in the UI.
+
+Render a form on it's own page if the form:
+
+- is complex and has many fields
+- will be used to complete an action that is very important
+
+### Section of a page
+
+A form may be rendered on a page alongside additional content to avoid taking users out of their current flow.
+
+Render a form alongside other content if:
+
+- the user may need to reference page content to help them fill out the form correctly
+- the form is not so complex that it deserves it's own page
+- it would be too disruptive to take a user away from the current page
+- the page has content that is related to the form
+
+### Dialog modal
+
+A form may be rendered in a dialog modal to keep the user on their current page while directing their focus to a form. The first input of the form should receive focus when the modal is opened.
+
+Render a form in a dialog modal if:
+
+- there are very few fields in the form
+- there is no logical place to render the form on the current page
+- the user won't need to reference content that will be covered by the dialog
+
+### Inline inputs
+
+In lieu of rendering an entire form, a single input can be used in place of the data it controls. Inputs would replace the data once the user has explicity toggled edit mode.
+
+Use an inline input if it would be too disruptive to show an entire form or take the user to a new page.
+
+## Structure and hierarchy
+
+Forms are designed with a consistent hierarchy and structure to make it easier for users to scan the inputs and submit their data.
+
+Every form will have at least one section, but form sections should always be used when a complex form can be broken down into distinct sections of related fields.
+
+For more information about the parts that make up a form, see the Storybook documentation for each component under the "Form structure" section.
+
+### Field order
+
+Order fields by their relative level of importance. The inputs that are most critical for the user to complete their task should come first.
+
+### Field grouping
+
+Fields that are closely related should be visually grouped together.
+
+For more information on grouped fields, see the Storybook documentation for the `FieldList` and `FieldGroup` components.
+
+### Form actions
+
+A form should contain at least two buttons: one to submit the form, and one to navigate away from the form without submitting any of the input.
```

Comment:
I don't know whether we want to make it a part of this guie, but I feel like it would fit here well:

In dcos-ui we sometimes show an "really discard?"-modal-Dialog when a user navigates away from a form that they have entered data into. 

I personally don't like that, as we could store that information to show it when a user next navigates to that form.

Reply:
> is not always expected nor desired, especially if user clicks cancel

fair point. i was more onto the "navigates away". 

e.g. on github i have the freedom to navigate awat to the "files" tab while authoring a PR or comment - and while i know i could open that in a new tab, i feel good every time i make use of that, because i get the feeling that their UI has me covered.

just my 2 cents.

Reply:
@danielone - we can add information about when to show a warning dialog.

When do you think we should show a warning dialog? When should we _not_ show a warning dialog?

Reply:
@mperrotti 
I think language may want to be from
"one to navigate away from the form without submitting any of the input."
to
"action to cancel the form and return the user to where accessed the form from."

Can we table the warning dialog part for now so not blocker to publishing? 
This way we can properly define the usage cases and also standard messaging for such.

Reply:
perhaps it is something you even "componentize" so dev can just hook up even if a ui design calls for it / part of standards, easy to include.. 

right now it seems like a good idea for any create/edit full page where data has been entered/edited where changes beyond displayname name field.